File: packages/java/anip-service/src/main/java/dev/anip/service/FailureRedaction.java
package dev.anip.service;

import java.util.HashMap;
import java.util.Map;

public final class FailureRedaction {
    private static final Map<String, String> GENERIC_MESSAGES = Map.ofEntries(
        Map.entry("scope_insufficient", "Insufficient scope for this capability"),
        Map.entry("invalid_token", "Authentication failed"),
        Map.entry("token_expired", "Token has expired"),
        Map.entry("purpose_mismatch", "Token purpose does not match this capability"),
        Map.entry("insufficient_authority", "Insufficient authority for this action"),
        Map.entry("unknown_capability", "Capability not found"),
        Map.entry("not_found", "Resource not found"),
        Map.entry("unavailable", "Service temporarily unavailable"),
        Map.entry("concurrent_lock", "Operation conflict"),
        Map.entry("internal_error", "Internal error"),
        Map.entry("streaming_not_supported", "Streaming not supported for this capability"),
        Map.entry("scope_escalation", "Scope escalation not permitted")
    );

    private FailureRedaction() {}

    @SuppressWarnings("unchecked")
    public static Map<String, Object> redact(Map<String, Object> failure, String level) {
        if ("full".equals(level)) {
            return new HashMap<>(failure);
        }
        var result = new HashMap<>(failure);
        if ("reduced".equals(level)) {
            String detail = (String) result.get("detail");
            if (detail != null && detail.length() > 200) {
                result.put("detail", detail.substring(0, 200));
            }
            if (result.get("resolution") instanceof Map) {
                var res = new HashMap<>((Map<String, Object>) result.get("resolution"));
                res.put("grantable_by", null);
                result.put("resolution", res);
            }
            return result;
        }
        // "redacted" mode
        String failType = (String) result.get("type");
        result.put("detail", GENERIC_MESSAGES.getOrDefault(failType, "Request failed"));
        if (result.get("resolution") instanceof Map) {
            var res = (Map<String, Object>) result.get("resolution");
            var redactedRes = new HashMap<String, Object>();
            redactedRes.put("action", res.get("action"));
            redactedRes.put("requires", null);
            redactedRes.put("grantable_by", null);
            redactedRes.put("estimated_availability", null);
            result.put("resolution", redactedRes);
        }
        return result;
    }
}

File: packages/csharp/src/Anip.Service/FailureRedaction.cs
namespace Anip.Service;

/// <summary>
/// Applies disclosure-level redaction to failure responses.
/// Rules: type, retry, resolution.action are never redacted.
/// "full": everything as-is.
/// "reduced": detail truncated to 200 chars, resolution.grantable_by nulled.
/// "redacted": detail replaced with generic message, resolution fields nulled except action.
/// </summary>
public static class FailureRedaction
{
    private static readonly Dictionary<string, string> GenericMessages = new()
    {
        ["scope_insufficient"] = "Insufficient scope for this capability",
        ["invalid_token"] = "Authentication failed",
        ["token_expired"] = "Token has expired",
        ["purpose_mismatch"] = "Token purpose does not match this capability",
        ["insufficient_authority"] = "Insufficient authority for this action",
        ["unknown_capability"] = "Capability not found",
        ["not_found"] = "Resource not found",
        ["unavailable"] = "Service temporarily unavailable",
        ["concurrent_lock"] = "Operation conflict",
        ["internal_error"] = "Internal error",
        ["streaming_not_supported"] = "Streaming not supported for this capability",
        ["scope_escalation"] = "Scope escalation not permitted",
    };

    /// <summary>
    /// Applies disclosure-level redaction to a failure response.
    /// The input is NOT mutated; a new dictionary is returned.
    /// </summary>
    public static Dictionary<string, object?> Redact(Dictionary<string, object?> failure, string level)
    {
        if (level == "full")
        {
            return CopyMap(failure);
        }

        var result = CopyMap(failure);

        if (level == "reduced")
        {
            if (result.TryGetValue("detail", out var detailObj) && detailObj is string detail && detail.Length > 200)
            {
                result["detail"] = detail[..200];
            }

            if (result.TryGetValue("resolution", out var resObj) && resObj is Dictionary<string, object?> res)
            {
                var resCopy = CopyMap(res);
                resCopy["grantable_by"] = null;
                result["resolution"] = resCopy;
            }

            return result;
        }

        // "redacted" mode
        var failType = result.TryGetValue("type", out var typeObj) ? typeObj as string : null;
        if (failType != null && GenericMessages.TryGetValue(failType, out var msg))
        {
            result["detail"] = msg;
        }
        else
        {
            result["detail"] = "Request failed";
        }

        if (result.TryGetValue("resolution", out var resObj2) && resObj2 is Dictionary<string, object?> res2)
        {
            var resCopy = new Dictionary<string, object?>
            {
                ["action"] = res2.TryGetValue("action", out var action) ? action : null,
                ["requires"] = null,
                ["grantable_by"] = null,
                ["estimated_availability"] = null,
            };
            result["resolution"] = resCopy;
        }

        return result;
    }

    private static Dictionary<string, object?> CopyMap(Dictionary<string, object?> source)
    {
        return new Dictionary<string, object?>(source);
    }
}
